rust/lance-linalg/src/distance/cosine.rs

```diff
@@ -89,7 +91,76 @@ where
 
 impl Cosine for BFloat16Type {}
 
-impl Cosine for Float16Type {}
+#[cfg(any(
+    all(target_os = "macos", target_feature = "neon"),
+    all(target_os = "linux", feature = "avx512fp16")
+))]
+mod kernel {
+    use super::*;
+
+    extern "C" {
+        pub fn dot_f16(ptr1: *const f16, ptr2: *const f16, len: u32) -> f32;
+        pub fn norm_l2_f16(ptr: *const f16, len: u32) -> f32;
+        pub fn cosine_fast_f16(ptr1: *const f16, ptr2: f32, ptr3: *const f16, len: u32) -> f32;
+    }
+}
+
+impl Cosine for Float16Type {
+    fn cosine(x: &[f16], y: &[f16]) -> f32 {
+        #[cfg(any(
+            all(target_os = "macos", target_feature = "neon"),
+            all(target_os = "linux", feature = "avx512fp16")
+        ))]
+        unsafe {
+            let x_norm: f32 = kernel::norm_l2_f16(x.as_ptr(), x.len() as u32);
+            kernel::cosine_fast_f16(x.as_ptr(), x_norm, y.as_ptr(), x.len() as u32)
+        }
+        #[cfg(not(any(
+            all(target_os = "macos", target_feature = "neon"),
+            all(target_os = "linux", feature = "avx512fp16")
+        )))]
+        {
+            let x_norm = norm_l2(x);
+            Self::cosine_fast(x, x_norm, y)
+        }
+    }
+
+    #[inline]
+    fn cosine_fast(x: &[f16], x_norm: f32, y: &[f16]) -> f32 {
+        #[cfg(any(
+            all(target_os = "macos", target_feature = "neon"),
+            all(target_os = "linux", feature = "avx512fp16")
+        ))]
+        unsafe {
+            kernel::cosine_fast_f16(x.as_ptr(), x_norm, y.as_ptr(), x.len() as u32)
+        }
+        #[cfg(not(any(
+            all(target_os = "macos", target_feature = "neon"),
+            all(target_os = "linux", feature = "avx512fp16")
+        )))]
+        {
+            cosine_scalar(x, x_norm, y)
+        }
+    }
+
+    #[inline]
+    fn cosine_with_norms(x: &[f16], x_norm: f32, y_norm: f32, y: &[f16]) -> f32 {
+        #[cfg(any(
+            all(target_os = "macos", target_feature = "neon"),
+            all(target_os = "linux", feature = "avx512fp16")
+        ))]
+        unsafe {
+            1.0 - kernel::dot_f16(x.as_ptr(), y.as_ptr(), x.len() as u32) / x_norm / y_norm
```

Comment:
do we need to handle zero `x_nrom` / `y_norm` here?

Reply:
Added `+ f32::epsilon()`.